batch/src/main/java/org/openmrs/analytics/FhirEtl.java

```diff
@@ -0,0 +1,161 @@
+    return new FhirSearchUtil(fhirStoreUtil, fhirContext);
+  }
+
+  static List<SearchSegmentDescriptor> createSegments(FhirEtlOptions options)
+      throws CannotProvideCoderException {
+    FhirSearchUtil fhirSearchUtil = creatFhirSearchUtil(options);
+    List<SearchSegmentDescriptor> segments = new ArrayList<>();
+    FhirContext fhirContext= FhirContext.forR4();
+    for (String resourceType : options.getSearchList().split(",")) {
+      String searchUrl = options.getServerUrl() + "/ws/fhir2/R4/" + resourceType;
+      Bundle searchBundle = fhirSearchUtil.searchForResource(searchUrl, options.getBatchSize(), 0,
+          options.getJsessionId(), true, fhirContext);
+      if (searchBundle == null) {
+        log.error("Cannot fetch resources for " + searchUrl);
+        throw new IllegalStateException("Cannot fetch resources for " + searchUrl);
+      }
+      int total = searchBundle.getTotal();
+      log.info(String.format("Number of resources for %s search is %d", resourceType, total));
+      if (searchBundle.getEntry().size() >= total) {
+        // No parallelism is needed in this case; we have all the resources already.
+        fhirSearchUtil.uploadBundleToCloud(searchBundle);
+      } else {
+        String baseNextUrl = fhirSearchUtil.findBaseSearchUrl(searchBundle);
+        for (int offset = 0; offset < total; offset += options.getBatchSize()) {
+          segments.add(SearchSegmentDescriptor.create(baseNextUrl, offset, options.getBatchSize(),
+              options.getJsessionId()));
+        }
+      }
+    }
+    log.info("Total number of segments is " + segments.size());
+    return segments;
+  }
+
+  static class FetchSearchPageFn extends DoFn<SearchSegmentDescriptor, DomainResource> {
+    private String gcpFhirStore;
+    private FhirContext fhirContext;
+    private FhirSearchUtil fhirSearchUtil;
+    private FhirStoreUtil fhirStoreUtil;
+
+    FetchSearchPageFn(String gcpFhirStore) {
+      this.gcpFhirStore = gcpFhirStore;
+    }
+
+    @Setup
+    public void Setup() {
+      this.fhirContext = FhirContext.forR4();
+      this.fhirStoreUtil = new FhirStoreUtil(gcpFhirStore);
+      this.fhirSearchUtil = new FhirSearchUtil(fhirStoreUtil, fhirContext);
+    }
+
+    @ProcessElement
+    public void ProcessElement(@Element SearchSegmentDescriptor segment,
+        OutputReceiver<DomainResource> out) {
+      Bundle pageBundle = fhirSearchUtil.searchForResource(segment.searchUrl(), segment.count(),
+          segment.pageOffset(), segment.jsessionId(), true, fhirContext);
+      fhirSearchUtil.uploadBundleToCloud(pageBundle);
+    }
+  }
+
+  static void runFhirFetch(FhirEtlOptions options) throws CannotProvideCoderException {
+    List<SearchSegmentDescriptor> segments = createSegments(options);
+    if (segments.isEmpty()) {
+      return;
+    }
+
+    Pipeline p = Pipeline.create(options);
+    p.apply(Create.of(segments)).apply(ParDo.of(new FetchSearchPageFn(options.getGcpFhirStore())));
```

Comment:
Does using the Beam API carry any additional benefits? Could we run this pipeline on different runners? 

I dabbled with a similar approach and tested it on a Flink runner here: https://github.com/pmanko/beam-local-sync/blob/master/src/main/java/org/itech/shr/beam/LocalFhirSync.java

Reply:
Yes there are a couple of benefits for using Beam one of which is being able to run on different "runners", e.g., Dataflow, Spark, Flink, local, etc. TBH, currently we are not benefiting much from the parallelization/distribution power of Beam/runners as the main bottleneck in this case is the single node MySQL DB and hence we are only running this pipeline through LocalRunner (and only benefit from multi-threading). But if in future we want to sync multiple OMRS instances using a single pipeline, being able to run on different runners can be very beneficial to us.

Haven't looked at the details of your implementation yet but speaking of Flink, I suppose your implementation is for the streaming case, correct? (BTW, this is another potential benefit of Beam where it is theoretically easier to share code between streaming and batch cases.) If yes, I am curious how you are taking data out of OpenMRS, is it through AtomFeed module as well?